File: app/models/plan.py
"""SQLAlchemy ORM models for user-managed application plans."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base


class Plan(Base):
    """Volunteer application plan owned by a user."""

    __tablename__ = "plan"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("user.id", ondelete="CASCADE"), index=True, nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    province: Mapped[str] = mapped_column(String(32), nullable=False)
    subject_type: Mapped[str] = mapped_column(String(32), nullable=False)
    score: Mapped[int] = mapped_column(Integer, nullable=False)
    rank: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[str] = mapped_column(String(32), default="draft", nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    items: Mapped[list["PlanItem"]] = relationship(
        back_populates="plan",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )


class PlanItem(Base):
    """School-major item inside a saved volunteer plan."""

    __tablename__ = "plan_item"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    plan_id: Mapped[int] = mapped_column(ForeignKey("plan.id", ondelete="CASCADE"), index=True, nullable=False)
    school_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    major_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    group_type: Mapped[str] = mapped_column(String(16), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    source_type: Mapped[str] = mapped_column(String(32), default="recommendation", nullable=False)
    recommend_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    risk_level: Mapped[str | None] = mapped_column(String(32), nullable=True)
    school_name_snapshot: Mapped[str] = mapped_column(String(128), nullable=False)
    major_name_snapshot: Mapped[str | None] = mapped_column(String(128), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    plan: Mapped[Plan] = relationship(back_populates="items")

File: frontend/src/types/plan.ts
export type PlanGroupType = "rush" | "stable" | "safe";
export type PlanSourceType = "manual" | "recommendation";

export interface PlanItemCreateRequest {
  school_id?: number | null;
  major_id?: number | null;
  school_name: string;
  major_name?: string | null;
  province?: string | null;
  city?: string | null;
  group_type: PlanGroupType;
  sort_order: number;
  source_type: PlanSourceType;
  recommend_reason?: string | null;
  risk_level?: string | null;
}

export interface CreatePlanRequest {
  name: string;
  province: string;
  subject_type: string;
  score: number;
  rank: number;
  status: string;
  items: PlanItemCreateRequest[];
}

export interface PlanSummary {
  id: number;
  user_id: number;
  name: string;
  province: string;
  subject_type: string;
  score: number;
  rank: number;
  status: string;
  created_at: string;
  updated_at: string;
  items_count: number;
}

export interface PlanItem {
  id: number;
  plan_id: number;
  school_id?: number | null;
  major_id?: number | null;
  school_name: string;
  major_name?: string | null;
  group_type: PlanGroupType;
  sort_order: number;
  source_type: PlanSourceType;
  recommend_reason?: string | null;
  risk_level?: string | null;
  created_at: string;
  updated_at: string;
}

export interface GroupedPlanItems {
  rush: PlanItem[];
  stable: PlanItem[];
  safe: PlanItem[];
}

export interface PlanDetail extends PlanSummary {
  items: PlanItem[];
  grouped_items: GroupedPlanItems;
}
